src/render/draw_symbol.js

```diff
@@ -124,36 +159,89 @@ function drawLayerSymbols(painter, sourceCache, layer, coords, isText, translate
             uglCoordMatrix = painter.translatePosMatrix(glCoordMatrix, tile, translate, translateAnchor, true);
 
+        const hasHalo = isSDF && layer.paint.get(isText ? 'text-halo-width' : 'icon-halo-width').constantOr(1) !== 0;
+
         let uniformValues;
         if (isSDF) {
-            const hasHalo = layer.paint.get(isText ? 'text-halo-width' : 'icon-halo-width').constantOr(1) !== 0;
 
             uniformValues = symbolSDFUniformValues(sizeData.functionType,
                 size, rotateInShader, pitchWithMap, painter, matrix,
                 uLabelPlaneMatrix, uglCoordMatrix, isText, texSize, true);
 
-            if (hasHalo) {
-                drawSymbolElements(buffers, layer, painter, program, depthMode, stencilMode, colorMode, uniformValues);
+            if (!sortByFeature) {
+                if (hasHalo) {
+                    drawSymbolElements(buffers, buffers.segments, layer, painter, program, depthMode, stencilMode, colorMode, uniformValues);
+                }
+                uniformValues['u_is_halo'] = 0;
             }
 
-            uniformValues['u_is_halo'] = 0;
-
         } else {
             uniformValues = symbolIconUniformValues(sizeData.functionType,
                 size, rotateInShader, pitchWithMap, painter, matrix,
                 uLabelPlaneMatrix, uglCoordMatrix, isText, texSize);
         }
 
-        drawSymbolElements(buffers, layer, painter, program, depthMode, stencilMode, colorMode, uniformValues);
+        if (sortByFeature) {
+            ((tileRenderState: any): Array<SymbolTileRenderState>).push({
+                buffers,
+                program,
+                depthMode,
+                uniformValues,
+                atlasTexture,
+                atlasInterpolation,
+                isSDF,
+                hasHalo
+            });
+        } else {
+            atlasTexture.bind(atlasInterpolation, gl.CLAMP_TO_EDGE);
+            drawSymbolElements(buffers, buffers.segments, layer, painter, program, depthMode, stencilMode, colorMode, uniformValues);
+        }
     }
 }
 
-function drawSymbolElements(buffers, layer, painter, program, depthMode, stencilMode, colorMode, uniformValues) {
+function drawSymbolElements(buffers, segments, layer, painter, program, depthMode, stencilMode, colorMode, uniformValues) {
     const context = painter.context;
     const gl = context.gl;
     program.draw(context, gl.TRIANGLES, depthMode, stencilMode, colorMode, CullFaceMode.disabled,
         uniformValues, layer.id, buffers.layoutVertexBuffer,
-        buffers.indexBuffer, buffers.segments, layer.paint,
+        buffers.indexBuffer, segments, layer.paint,
         painter.transform.zoom, buffers.programConfigurations.get(layer.id),
         buffers.dynamicLayoutVertexBuffer, buffers.opacityVertexBuffer);
 }
+
+function drawSymbolsSorted(painter: Painter, renderData: Array<SymbolTileRenderState>, layer, colorMode, stencilMode) {
```

Comment:
Can we refactor this so it's "prepare draw state per tile, then draw once per segment" whether we're sorting or not? I think it would be easier to follow that way -- we wouldn't have to deal with the ambiguous naming that `drawLayerSymbols` is sometimes a preparation step and sometimes an actual draw step.

Reply:
I think my thinking here was to avoid the extra overhead for symbol layers that don't need this but I should probably benchmark to see if this is valid. It looks like it should be pretty cheap to do what you are suggesting.

Reply:
@ChrisLoer I've pushed a commit to always use the preparation step. Can you review?